src/network_service/sending_targets_cache.rs

```diff
@@ -106,28 +100,58 @@ impl SendingTargetsCache {
     }
 }
 
-fn set_target_failed(targets: &mut Vec<(ConnectionInfo, TargetState)>, target: &SocketAddr) {
+/// Info about failed target.
+pub struct TargetFailed {
+    /// Next target to try to resend the message to, if any.
+    pub next: Option<ConnectionInfo>,
+    /// Whether the failed target should be considered as lost - that is - whether it failed more
+    /// than maximum allowed number of times.
+    pub lost: bool,
+}
+
+// Marks the target as failed and returns whether it is lost.
+fn set_target_failed(
+    targets: &mut Vec<(ConnectionInfo, TargetState)>,
+    target: &SocketAddr,
+) -> bool {
     let (index, state) = if let Some((index, (_, state))) = targets
         .iter_mut()
         .enumerate()
         .find(|(_, (info, _))| info.peer_addr == *target)
     {
         (index, state)
     } else {
-        return;
+        return false;
     };
 
     match *state {
         TargetState::Failed(_) => {
             log_or_panic!(LogLevel::Error, "Got a failure from a failed target!");
             let _ = targets.remove(index);
+            true
         }
         TargetState::Sending(x) => {
             if x < MAX_RESENDS {
                 *state = TargetState::Failed(x + 1);
+                false
             } else {
                 let _ = targets.remove(index);
+                true
```

Comment:
I'm wondering if it would make more sense to return the final result on success or fail i.e if we remove the collection we send it back or the item that failed max number of times. (for both the success part and fail part)? not sure.